haddock-api/src/Haddock.hs

```diff
@@ -504,51 +499,79 @@ unsetPatternMatchWarnings dflags =
+      -- When in the GHC tree, we should be able to locate the "lib" folder
+      -- based on the location of the current executable.
+      base_dir <- getBaseDir      -- Provided by GHC
+      let res_dirs = [ d | Just d <- [base_dir] ] ++
+
 #else
-      -- if data directory does not exist we are probably
-      -- invoking from either ./haddock-api or ./
-      let res_dirs = [ getDataDir -- provided by Cabal
-                     , pure "resources"
-                     , pure "haddock-api/resources"
-                     ]
 
-          check get_path = do
-            p <- get_path
-            exists <- doesDirectoryExist p
-            pure $ if exists then Just p else Nothing
+      -- When Haddock was installed by @cabal@, the resources (which are listed
+      -- under @data-files@ in the Cabal file) will have been copied to a
+      -- special directory.
+      data_dir <- getDataDir      -- Provided by Cabal
+      let res_dirs = [ data_dir ] ++
 
-      dirs <- mapM check res_dirs
-      case [p | Just p <- dirs] of
-        (p : _) -> return p
-        _       -> die "Haddock's resource directory does not exist!\n"
 #endif
-    fs -> return (last fs)
 
+      -- When Haddock is built locally (eg. regular @cabal new-build@), the data
+      -- directory does not exist and we are probably invoking from either
+      -- @./haddock-api@ or @./@
+                     [ "resources"
+                     , "haddock-api/resources"
+                     ]
 
-getGhcDirs :: [Flag] -> IO (String, String)
+      res_dir <- check res_dirs
+      case res_dir of
+        Just p -> return p
+        _      -> die "Haddock's resource directory does not exist!\n"
+
+    fs -> return (last fs)
+  where
+    -- Pick the first path that corresponds to a directory that exists
+    check :: [FilePath] -> IO (Maybe FilePath)
+    check [] = pure Nothing
+    check (path : other_paths) = do
+      exists <- doesDirectoryExist path
+      if exists then pure (Just path) else check other_paths
+
+-- | Find the @lib@ directory for GHC and the path to @ghc@
+getGhcDirs :: [Flag] -> IO (Maybe FilePath, Maybe FilePath)
 getGhcDirs flags = do
-  case [ dir | Flag_GhcLibDir dir <- flags ] of
-    [] -> do
+
 #ifdef IN_GHC_TREE
-      libDir <- getInTreeDir
-      return (ghcPath, libDir)
+  base_dir <- getBaseDir
+  let ghc_path = Nothing
 #else
-      return (ghcPath, GhcPaths.libdir)
+  let base_dir = Just GhcPaths.libdir
+      ghc_path = Just GhcPaths.ghc
 #endif
-    xs -> return (ghcPath, last xs)
-  where
+
+  -- If the user explicitly specifies a lib dir, use that
+  let ghc_dir = case [ dir | Flag_GhcLibDir dir <- flags ] of
+                  [] -> base_dir
+                  xs -> Just (last xs)
+
+  pure (ghc_path, ghc_dir)
+
+
 #ifdef IN_GHC_TREE
-    ghcPath = "not available"
-#else
-    ghcPath = GhcPaths.ghc
-#endif
 
+-- | See 'getBaseDir' in "SysTools.BaseDir"
+getBaseDir :: IO (Maybe FilePath)
+getBaseDir = do
+  exec_path <- getExecutablePath
```

Comment:
@hvr Is this safe across all platforms?

The previous version of this didn't use `getExecutablePath`. Additionally, I've seen other code where uses of `getExecutablePath` are limited via CPP to only some platforms (darwin, windows, and linux I think), but I can't find any justification for this.

Reply:
Couple points from discord convo, and my resolution of them:

  * __Windows might be relying on this working, due to some issues preventing wrapper scripts.__
    I've double checked and `getExecutablePath` in `System.Environment.ExecutablePath` is just a slightly better take on the old `getExecDir` (same call to `GetModuleFileNameW`), so I doubt we'll regress here.
  * __`getExecutablePath` isn't reliable.__
    Looking again at the implementation, we should be ready to catch  exceptions throw by `getBaseDir` and return `Nothing` in those cases.

  * __Even if it returns something, we should validate it.__
    Checking for the existence of the `lib` directory seems like a best effort here. Good enough for now.